cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -273,6 +328,29 @@ struct Atan {
   }
 };
 
+struct Atanh {
+  template <typename T, typename Arg0>
+  static enable_if_floating_value<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE((val < -1.0 || val > 1.0))) {
```

Comment:
No, it's actually a domain error:

```
Cell In[24], line 1
----> 1 math.atanh(-1.0)

ValueError: math domain error

In [25]: math.atanh(1.0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[25], line 1
----> 1 math.atanh(1.0)

ValueError: math domain error
```

Reply:
```suggestion
    if (ARROW_PREDICT_FALSE((val <= -1.0 || val => 1.0))) {
```

Reply:
I actually went back and forth on this. In *python* they raise a domain error. In the C++ stdlib, they return ± infinity. So the code as is returns ± infinity and an error is raised (which I realize I didn't clear....). See [here](https://en.cppreference.com/w/cpp/numeric/math/atanh).

Which library this PR should be consistent with

Reply:
C++ docs also say `If the argument is ±1, a pole error occurs.` and `if the argument is ±1, ±∞ is returned and [FE_DIVBYZERO] is raised.

They have different names for the errors (range/pole/FE_DIVBYZERO) but that doesn't change the fact that Mathematically, `atanh` isn't defined at the -1.0 and 1.0 points.

The function **converges to** $\pm\infty$ as $x$ approaches the extremes of the range which makes returning `-Inf/+Inf` more reasonable than returning `NaN` at these extremes, but it's not defined at these points per se.

$$\lim_{x \to 1.0}{tanh^{-1}(x)} = \infty$$
$$\lim_{x \to -1.0}{tanh^{-1}(x)} = -\infty$$

So I think you can drop my suggestion on the unchecked version (this one) and let the implementation return `-Inf/+Inf` instead of `NaN` at these extremes, but the `atanh_checked` variation should fail at `-1.0/+1.0` instead of returning the infinite values.

Docs should be updated accordingly to describe this behavior.

Reply:
FWIW, a common approach when implementing Acero functions in cases like this has been to add options to allow the user can pick what behavior they want. It might be worth doing that here if it doesn't add excessive complexity or inefficiency.

Reply:
The logarithms definition of `atanh` makes the issue more apparent:

$$tanh^{-1}(x) = \frac{1}{2}\ln\frac{1 + x}{1 - x}$$

If x is 1, you get a division by 0. If x is -1 you get $\ln(0)$ which is undefined (converges to `-Inf`, but not defined).

Reply:
Oh totally! I honestly think both implementations can be rigorously justified and defended.

For instance, I could *define* atanh as the inverse of tanh. And I could *define* f(oo) = lim_{x -> oo} f(x). And then atanh(1) is clearly oo.

Just a matter of which convention the library wants to follow 😀

Reply:
Notably, C++ defines [tanh(oo) = 1](https://en.cppreference.com/w/cpp/numeric/math/tanh)

Reply:
Thanks for digging in all. I think the direction this conversion went makes sense (unchecked version returns -Inf/Inf, checked version raises). I was checking these in R and SQLite and it looks like Python is sort of the odd man out here.

Reply:
Sounds good. Change made.